File: RUBY/src/Weapon.rb
#encoding: UTF-8

require_relative 'WeaponType'
require_relative 'WeaponToUI'

module Deepspace

	class Weapon

		def initialize(n,t,u)
			@name = n
			@type = t
			@uses = u
		end
		
		def self.newCopy(other)
			new(other.name, other.type, other.uses)
		end
		
		def getUIversion
			WeaponToUI.new(self)
		end

		attr_reader :name
		attr_reader :uses
		attr_reader :type

		def power
			@type.power()
		end

		def useIt
			if uses > 0
				@uses -= 1
				power()
			else
				1.0
			end
		end
		
		def to_s
			if (@type == WeaponType::LASER)
				type = "LASER"
			else 
				if (@type == WeaponType::MISSILE)
					type = "MISSILE"
				else
					type = "PLASMA"
				end
			end
			
			return "Name: " + @name + ", Type: " + type +", Power: #{@type.power}, Uses: #{@uses}"
		end
		
	end	
end

File: Practica5/Marcianitos/src/deepspace/Weapon.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

/**
 *
 * @author mario
 */
public class Weapon implements CombatElement{
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon (String the_name, WeaponType the_type, int the_uses){
        this.name = the_name;
        this.type = the_type;
        this.uses = the_uses;
    }
    
    Weapon (Weapon w){
        this.name = w.name;
        this.type = w.type;
        this.uses = w.uses;
    }
    
    WeaponToUI getUIversion() {
        return new WeaponToUI(this);
    }
    
    public WeaponType getType (){
        return (this.type);
    }
    
    // ns si esto está bien pq no tiene ninguna utilidad poner exactamente lo mismo
    // pero q haya una interfaz
    @Override
    public int getUses(){
        return (this.uses);
    }
    
    
    public float power (){
        return (type.getPower());
    }
    
    // lo mismo q el getUses, no creo q esté bien
    @Override
    public float useIt(){
        float retorno = 1;
        
        if (this.uses > 0){
            --uses;
            retorno = this.power();
        }
        
        return (retorno);
    }
    
    public String toString(){
        
        String salida;
        
        salida = "Name=" + name + ", Type=" + type + ", Uses=" + uses;
        
        return salida;
    }
}
